### PR TITLE
[BPK-2049] add white cta

### PR DESCRIPTION
**Pull request by georgegillams**
I think this is now aligned to Android...

See @amolgupta's PR for reference:
https://github.com/Skyscanner/backpack-android/pull/127#issue-233129394

Still todo:
 - [x] Rename `White` to `Outline`
 - [x] Clean up internal var naming

screenshot 2018-11-29 at 14 30 27

### CODE REVIEW

File: Backpack/Button/Classes/BPKButton.m [hunk 1]
```diff
@@ -291,6 +293,7 @@ - (UIEdgeInsets)contentEdgeInsetsForStyle:(BPKButtonStyle)style size:(BPKButtonS
 
 - (void)updateBackgroundAndStyle {
     UIColor *highlightedWhite = [UIColor colorWithRed:0.871 green:0.867 blue:0.878 alpha:1];
+    UIColor *highlightedWhite2 = [UIColor colorWithRed:0 green:0 blue:0 alpha:0.15];
```
Comment (georgegillams) [inline Backpack/Button/Classes/BPKButton.m]: Obviously needs renaming!

File: Backpack/Button/Classes/BPKButton.m [hunk 2]
```diff
@@ -489,18 +509,30 @@ - (void)setLinkStyleWithColor:(UIColor *)color {
...
 }
 
 + (UIColor *)highlightedWhite {
     // white overlayed with gray900 at 15%
     // TODO: Add a method to blend programatically via BPKColor
-    return [UIColor colorWithRed:0.871 green:0.867 blue:0.878 alpha:1];
+    return [UIColor colorWithRed:1 green:1 blue:1 alpha:0.15];
+}
+
++ (UIColor *)highlightedWhite2 {
```
Comment (georgegillams) [inline Backpack/Button/Classes/BPKButton.m]: Obviously needs renaming!

File: Example/Backpack/View Controllers/BPKButtonsViewController.m [hunk 3]
```diff
@@ -20,9 +20,12 @@
 
 #import <Backpack/Icon.h>
 #import <Backpack/Color.h>
+#import <Backpack/Label.h>
 
 NS_ASSUME_NONNULL_BEGIN
 @interface BPKButtonsViewController ()
+@property (strong, nonatomic) IBOutletCollection(UIView) NSArray *contentViews;
```
Comment (georgegillams) [inline Example/Backpack/View Controllers/BPKButtonsViewController.m]: I appreciate "contentViews" is a bit of a confusing term as the norm is to have a single `contentView`. Maybe there's a better name?

File: Backpack/Button/Classes/BPKButton.m [hunk 4]
```diff
@@ -503,6 +532,12 @@ + (UIColor *)highlightedWhite {
     return [UIColor colorWithRed:0.871 green:0.867 blue:0.878 alpha:1];
 }
 
++ (UIColor *)highlightedWhite2 {
+    // white overlayed with gray900 at 15%
+    // TODO: Add a method to blend programatically via BPKColor
+    return [UIColor colorWithRed:1 green:1 blue:1 alpha:0.75];
```
Comment (shaundon) [inline Backpack/Button/Classes/BPKButton.m]: I recently discovered [`colorWithAlphaComponent`](https://developer.apple.com/documentation/uikit/uicolor/1621922-colorwithalphacomponent?language=objc) that might be useful here:

```
return [BPKColor.gray900 colorWithAlphaComponent:0.75];

```

Comment (georgegillams) [reply Backpack/Button/Classes/BPKButton.m]: Good shout - I won't change this just yes as the exact color still needs to be agreed 👍

File: Backpack/Button/Classes/BPKButton.m [hunk 5]
```diff
@@ -474,7 +490,11 @@ - (void)setFilledStyleWithNormalBackgroundColorGradientOnTop:(UIColor *)normalCo
...
 
 - (void)setBorderedStyleWithColor:(UIColor *)color {
-    self.gradientLayer.gradient = [self gradientWithSingleColor:BPKColor.white];
+    [self setBorderedStyleWithColor:color withGradientColor:BPKColor.white];
+}
+
+- (void)setBorderedStyleWithColor:(UIColor *)color withGradientColor:(UIColor *)gradientColor {
+    self.gradientLayer.gradient = [self gradientWithSingleColor:gradientColor];
 
     UIColor *borderColor = BPKColor.gray100;
```
Comment (georgegillams) [inline Backpack/Button/Classes/BPKButton.m]: This should be setting the border white, but instead is hard-coded grey 100!

File: Backpack/Button/Classes/BPKButton.m (hunk 4 above)
Comment (georgegillams) [reply Backpack/Button/Classes/BPKButton.m]: Done now 😃

File: Backpack/Button/Classes/BPKButton.m (hunk 5 above)
Comment (georgegillams) [reply Backpack/Button/Classes/BPKButton.m]: Fixed 👍

File: Backpack/Button/Classes/BPKButton.m [hunk 6]
```diff
@@ -296,6 +296,9 @@ - (void)updateBackgroundAndStyle {
     UIColor *highlightedWhite2 = [UIColor colorWithRed:0 green:0 blue:0 alpha:0.15];
 
     if (self.isEnabled) {
+
+        // We need this here so that if the button was disabled, and is now enabled, opacity is reset.
+        self.layer.opacity = 1;
```
Comment (georgegillams) [inline Backpack/Button/Classes/BPKButton.m]: I don't know whether this belongs elsewhere?

**Comment by shaundon** [comment]
Just recording so we don't forget:

* The border should darken when the button is pressed

File: Backpack/Button/Classes/BPKButton.m [hunk 7]
```diff
@@ -301,21 +307,29 @@ - (void)updateBackgroundAndStyle {
...
+                [self setBorderedStyleWithColor:BPKColor.gray100];
                 if (self.isHighlighted) {
                     self.gradientLayer.gradient = [self gradientWithSingleColor:highlightedWhite];
                     [self.gradientLayer setNeedsDisplay];
                 }
                 break;
             }
             case BPKButtonStyleDestructive: {
-                [self setBorderedStyleWithColor:BPKColor.red500];
+                [self setBorderedStyleWithColor:BPKColor.gray100];
```
[the remark this answers is not in the document]
Comment (georgegillams) [reply Backpack/Button/Classes/BPKButton.m]: Previously the function `setBorderedStyleWithColor` took an argument `color` but applied gray100 regardless. I've fixed the function to actually use the supplied color, but so as not to break these I've had to change them :)

File: Backpack/Button/Classes/BPKButton.m [hunk 8]
```diff
@@ -301,21 +307,30 @@ - (void)updateBackgroundAndStyle {
...
                 if (self.isHighlighted) {
                     self.gradientLayer.gradient = [self gradientWithSingleColor:highlightedWhite];
                     [self.gradientLayer setNeedsDisplay];
                 }
                 break;
             }
+            case BPKButtonStyleOutline: {
+                [self setBorderedStyleWithColor:BPKColor.white withGradientColor:BPKColor.clear];
+                if (self.isHighlighted) {
```
Comment (shaundon) [inline Backpack/Button/Classes/BPKButton.m]: `setBorderedStyleWithColor` is getting called twice here if `isHighlighted == true`, might be more efficient to do the first one in an `else` block

File: Backpack/Button/Classes/BPKButton.m [hunk 9]
```diff
@@ -503,6 +528,12 @@ + (UIColor *)highlightedWhite {
     return [UIColor colorWithRed:0.871 green:0.867 blue:0.878 alpha:1];
 }
 
++ (UIColor *)highlightedOutline {
+    // white overlayed with gray900 at 15%
```
Comment (shaundon) [inline Backpack/Button/Classes/BPKButton.m]:
```suggestion
    // white overlaid with gray900 at 15%
```

File: Backpack/Button/Classes/BPKButton.m [hunk 10]
```diff
@@ -503,6 +528,12 @@ + (UIColor *)highlightedWhite {
...
+    // TODO: Add a method to blend programatically via BPKColor
+    return [BPKColor.white colorWithAlphaComponent:0.8];
```
Comment (shaundon) [inline Backpack/Button/Classes/BPKButton.m]: This comment says it's white overlaid with gray at 15%, but the code seems to just be white at 80% opacity?

File: UNRELEASED.md [hunk 11]
```diff
@@ -1 +1,7 @@
-# Unreleased
\ No newline at end of file
+# Unreleased
+
+**Added:**
+
+- Backpack/Button
+  - New `white` configuration which renders just an outline and the text. Designed for use on more colourful backgrounds.
```
Comment (shaundon) [inline UNRELEASED.md]:
```suggestion
  - New `outline` button type which renders just an outline and the text in white. Intended for use on coloured backgrounds.
```

File: Backpack/Button/Classes/BPKButton.m (hunk 8 above)
Comment (georgegillams) [reply Backpack/Button/Classes/BPKButton.m]: Nah :)

Comment (shaundon) [reply Backpack/Button/Classes/BPKButton.m]: anigif_sub-buzz-19873-1543319955-12

**Pull request merged**